### PR TITLE
feat: add removed_indices to CreateIndex transaction operation

### PR DESCRIPTION
**Pull request by eddyxu**

### CODE REVIEW

File: rust/lance/src/dataset/transaction.rs [hunk 1]
```diff
@@ -250,10 +259,12 @@ impl Transaction {
                 &other.operation,
                 Operation::Overwrite { .. } | Operation::Restore { .. }
             ),
-            Operation::CreateIndex { .. } => match &other.operation {
+            Operation::CreateIndex { .. } | Operation::ModifyIndex { .. } => match &other.operation
+            {
                 Operation::Append { .. } => false,
                 // Indices are identified by UUIDs, so they shouldn't conflict.
                 Operation::CreateIndex { .. } => false,
+                Operation::ModifyIndex { .. } => false,
```
Comment (chebbyChefNEQ) [inline rust/lance/src/dataset/transaction.rs:267]: if two ops are mofifying the same index they should conflict right?

Comment (eddyxu) [reply rust/lance/src/dataset/transaction.rs:267]: txn log is ordered , right? it means modifying index then recreate an index over it?

File: rust/lance/src/dataset/transaction.rs [hunk 2]
```diff
@@ -423,6 +436,18 @@ impl Transaction {
...
             }
+            Operation::ModifyIndex {
+                new_indices,
+                old_indices,
+            } => {
+                final_fragments.extend(maybe_existing_fragments?.clone());
+                final_indices.retain(|existing_index| {
+                    !old_indices
+                        .iter()
+                        .any(|old_index| old_index.uuid == existing_index.uuid)
```
Comment (chebbyChefNEQ) [inline rust/lance/src/dataset/transaction.rs:447]: would it make more sense to call these `removed_indices` and `added_indices`?

Maybe we can roll this up into `CreateIndex` op or vice versa?

Comment (eddyxu) [reply rust/lance/src/dataset/transaction.rs:447]: sg.

Comment (eddyxu) [reply rust/lance/src/dataset/transaction.rs:447]: > Maybe we can roll this up into CreateIndex op or vice versa?

what do you mean?

File: rust/lance/src/dataset/transaction.rs (hunk 1 above)
Comment (westonpace) [reply rust/lance/src/dataset/transaction.rs:267]: I think I agree with @eddyxu .  We do the same thing, for example, with overwrite / restore.  Both operations succeed even though one of them is immediately obsoleted.

File: rust/lance/src/dataset/transaction.rs (hunk 2 above)
Comment (westonpace) [reply rust/lance/src/dataset/transaction.rs:447]: I think the argument is that we only need a single operation that is adding and/or removing indices.  A `CreateIndex` could be a `ModifyIndex` where `old_indices` is an empty vec.  This way we have fewer operations to worry over.

**Review by westonpace — CHANGES_REQUESTED** [review]
In `conflicts_with` you missed a case for `Operation::Delete { .. } => match &other.operation`.  You should add `ModifyIndex => false` there as well.

Can you add a modify index case to the `test_conflicts` unit test?

File: rust/lance/src/dataset/transaction.rs (hunk 1 above)
Comment (wjones127) [reply rust/lance/src/dataset/transaction.rs:267]: I think modify index should conflict with modify index, if they modify the same index. I think it is compatible with CreateIndex (unless CreateIndex is overwriting the index it modifies).

These are transactions that happened at the same time, so you can't commit modified files if the files those are based on have been modified concurrently.

File: rust/lance/src/dataset/transaction.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance/src/dataset/transaction.rs:447]: We can not change `CreateIndex`'s name now, only add new fields?  

However, `CreateIndex { old_indices, new_indices }` does sound right?

Or do we want to deprecate `CreateIndex` and only have `ModifyIndex`?

Comment (wjones127) [reply rust/lance/src/dataset/transaction.rs:447]: `CreateIndex { old_indices, new_indices }` or `CreateIndex { removed_indices, new_indices }` sounds good to me.

File: rust/lance/src/dataset/transaction.rs [hunk 3]
```diff
@@ -414,12 +416,18 @@ impl Transaction {
...
+            Operation::CreateIndex {
+                new_indices,
+                removed_indices,
+            } => {
                 final_fragments.extend(maybe_existing_fragments?.clone());
                 final_indices.retain(|existing_index| {
                     !new_indices
                         .iter()
                         .any(|new_index| new_index.name == existing_index.name)
+                        || removed_indices
```
Comment (westonpace) [inline rust/lance/src/dataset/transaction.rs:428]: Should this be `&&`?

**Review by westonpace — COMMENTED** [review]
One last question, otherwise this is good

File: rust/lance/src/dataset/transaction.rs (hunk 3 above)
Comment (eddyxu) [reply rust/lance/src/dataset/transaction.rs:428]: OH that's right.

**Review by westonpace — APPROVED** [review]

**Pull request merged**